Configuration

MattSurabian opened the pull request:
Read configuration values via the command line or from a .secrets config file. Includes several configuration flags that will be leveraged in future PRs for TLS and NaCl encryption.

Method provided for `key-gen` command is only for user convenience and has no logic in place to actually generate keys. If a configuration file is added but no NaCl keys can actually be found the user is instructed to run `key-gen`.

MattSurabian commented:
@tafsiri or @mzgoddard if you get a chance could you review this PR?

tafsiri commented:
@MattSurabian an overall comment, I think it could use with some more documentation. An overall module comment for the various files and docs for at least the exported functions. Even smaller functions (but not the getters) could benefit, for example how AbsPath differs from http://golang.org/pkg/path/filepath/#Abs. Overall i think that would help with people being able to get into the project as well as the coming back to it in 3 months kinda thing.

tafsiri commented on client/command/return_codes.go:
```diff
@@ -0,0 +1,9 @@
+package command
+
+// Return codes to be used by command implementations and tests
+const (
```
nice :)

tafsiri commented on client/commands.go:
```diff
@@ -10,16 +10,19 @@ import (
+	}
 
 	Commands = map[string]cli.CommandFactory{
-		"get": func() (cli.Command, error) {
```
i'm not entirely sure why get and put are being removed as commands but i suspect that has to do with overall changes to project structure?

MattSurabian replied:
Yeah these were some initial commands added earlier in the project as place holders but now that I'm adding real ones I just removed the ones that don't actually do anything.

tafsiri commented on shared/cliUtils/config.go:
```diff
@@ -0,0 +1,278 @@
+)
+
+// In the event the config flag isn't passed this is the filename that will be searched for
+var defaultConfigFileName = ".secrets"
```
maybe consider const for this?

MattSurabian replied:
Definitely.

MattSurabian commented:
Good call @tafsiri I will update with method docs and high level module comments

tafsiri commented on shared/cliUtils/config.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func promptForAndPersistFlagValues(f *flag.Flag) {
+	if f.Name == "config" {
```
maybe some docs about why this check/case exists

MattSurabian replied:
Yeah I think what might be ideal is using something called flag sets to prevent these kinds of annoying mystery checks. Need to read up on that a little.

MattSurabian replied:
Unfortunately flag sets don't work the way I thought they did so I think I'll just need documentation about this :-(

MattSurabian commented:
Rebased away the removal of get/put since there's no harm in having commands that don't do anything yet in the codebase.

Addressed documentation issues per @tafsiri's recommendations.

MattSurabian commented:
Reviewed in person with @tafsiri and implemented his feedback (he legit gave me a :+1: IRL so :racehorse: . Hurray for configuration.
